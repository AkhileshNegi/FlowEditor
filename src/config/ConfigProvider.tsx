import * as React from 'react';
import { Endpoints, FlowEditorConfig } from '~/flowTypes';

export const fakePropType: any = (): any => null;
fakePropType.isRequired = (): any => null;

interface ConfigProviderProps {
    config: FlowEditorConfig;
    children: any;
}

export interface ConfigProviderContext {
    endpoints: Endpoints;
    flow: string;
    debug: boolean;
    showDownload: boolean;
}

// ----------------------------------------------------------------------------------------------

export const SINGLE_CHILD_ERROR = 'ConfigProvider expects only one child component.';
export const VALID_CHILD_ERROR =
    'ConfigProvider expects a valid React element: https://reactjs.org/docs/react-api.html#isvalidelement';

export default class ConfigProvider extends React.Component<ConfigProviderProps> {
    public static childContextTypes = {
        assetService: fakePropType,
        endpoints: fakePropType,
        flow: fakePropType,
        debug: fakePropType,
        showDownload: fakePropType
    };

    constructor(props: ConfigProviderProps) {
        super(props);

        if (React.Children.count(props.children) > 1) {
            throw new Error(SINGLE_CHILD_ERROR);
        } else if (!React.isValidElement(props.children)) {
            throw new Error(VALID_CHILD_ERROR);
        }
    }

    public getChildContext(): ConfigProviderContext {
        return {
            endpoints: this.props.config.endpoints,
            flow: this.props.config.flow,
            debug: this.props.config.debug,
            showDownload: this.props.config.showDownload
        };
    }

    public render(): JSX.Element {
        return React.Children.only(this.props.children);
    }
}
