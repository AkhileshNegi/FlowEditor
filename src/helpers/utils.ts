import { ComponentClass, SFC, ReactElement } from 'react';
import { ShallowWrapper, ReactWrapper } from 'enzyme';
import { FlowDefinition } from '../flowTypes';

const request = require('sync-request');

const XRegExp = require('xregexp');

const SNAKED_CHARS = XRegExp('[^\\p{Letter}\\d]+');

interface BoolMap {
    [key: string]: boolean;
}

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {object} A map of each string
 */
export function toBoolMap(array: string[]): BoolMap {
    return array.reduce((map: BoolMap, item: string) => {
        map[item] = true;
        return map;
    }, {});
}

/**
 * Inserts commas into numbers where appropriate for better readability
 * @param {number} value - A number, e.g. 10000
 * @returns {string} A comma-separated string, e.g. 10,000
 */
export function addCommas(value: number): string {
    return value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export function snakify(value: string): string {
    return XRegExp.replace(value.toLowerCase().trim(), SNAKED_CHARS, '_', 'all');
}

/**
 * Returns a React component's name so we can attach it to a HOC's displayName property and view it in RDT
 * @param {ComponentClass | SFC} Component - A React component
 * @returns {string} The component's name
 */
export function getDisplayName(HOCName: string, Component: ComponentClass | SFC): string {
    const ComponentDisplayName = Component.displayName || Component.name || 'Component';
    return `${HOCName}(${ComponentDisplayName})`;
}

function getFlow(name: string): FlowDefinition {
    const definition = request('GET', 'base/test_flows/' + name + '.json').getBody();
    return JSON.parse(definition) as FlowDefinition;
}

export function getFavorites(): FlowDefinition {
    return getFlow('favorites');
}

export function getTest(): FlowDefinition {
    return getFlow('a4f64f1b-85bc-477e-b706-de313a022979');
}

export function dump(object: any) {
    console.log(JSON.stringify(object, null, 1));
}

export const DATA_SPEC_ATTRIBUTE_NAME: string = 'data-spec';

/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} matching DOM components
 */
export function getSpecWrapper(
    componentWrapper: ReactWrapper | ShallowWrapper,
    specName: string
): ShallowWrapper | ReactWrapper {
    return componentWrapper.find(`[${DATA_SPEC_ATTRIBUTE_NAME}="${specName}"]`);
}

/**
 * Returns true if a given UUID matches v4 format
 * @param {string} uuid - A version 4 UUID (no braces, uppercase OK)
 * @returns {boolean}
 */
export function validUUID(uuid: string): boolean {
    const regex = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;
    return regex.test(uuid);
}
