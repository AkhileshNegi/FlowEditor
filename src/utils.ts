const XRegExp = require('xregexp');

const SNAKED_CHARS = XRegExp('[^\\p{Letter}\\d]+');

interface BoolMap {
    [key: string]: boolean;
}

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {BoolMap} A map of each string
 */
export function toBoolMap(array: string[]): BoolMap {
    return array.reduce((map: BoolMap, item: string) => {
        map[item] = true;
        return map;
    }, {});
}

export function addCommas(value: number): string {
    return value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export function snakify(value: string): string {
    return XRegExp.replace(value.toLowerCase(), SNAKED_CHARS, '_', 'all');
}
