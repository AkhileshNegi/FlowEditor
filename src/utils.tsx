/** Turns a string array into a bool map */
export function toBoolMap(array: string[]): {[item: string]: boolean} {
    var map: {[item: string]: boolean} = {};
    for (let item of array) {
        map[item] = true;
    }
    return map;
}