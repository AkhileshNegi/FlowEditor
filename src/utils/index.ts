import { ComponentClass, SFC, ReactElement } from 'react';
import { ShallowWrapper, ReactWrapper } from 'enzyme';
import { FlowDefinition, Node, Languages, SwitchRouter, AnyAction } from '../flowTypes';
import { DragPoint } from '../component/Node';
import { PendingConnections, Components } from '../redux';
import { Language } from '../component/LanguageSelector';
import Localization, { LocalizedObject } from '../services/Localization';
import { NodeEditorProps } from '../component/NodeEditor';

const SNAKED_CHARS = /\s+(?=\S)/g;
const GRID_SIZE = 20;
export const V4_UUID = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;

interface BoolMap {
    [key: string]: boolean;
}

interface Bounds {
    left: number;
    top: number;
    right: number;
    bottom: number;
}

/**
 * Adjusts the left and top offsets to a grid
 * @param left horizontal offset
 * @param top vertical offset
 */
export const snapToGrid = (left: number, top: number): { left: number; top: number } => {
    // Adjust our ghost into the grid
    let leftAdjust = left % GRID_SIZE;
    let topAdjust = top % GRID_SIZE;

    if (leftAdjust > GRID_SIZE / 3) {
        leftAdjust = GRID_SIZE - leftAdjust;
    } else {
        leftAdjust = leftAdjust * -1;
    }

    if (topAdjust > GRID_SIZE / 3) {
        topAdjust = GRID_SIZE - topAdjust;
    } else {
        topAdjust = topAdjust * -1;
    }

    return {
        left: Math.max(left + leftAdjust, GRID_SIZE * 2),
        top: Math.max(top + topAdjust, GRID_SIZE * 4)
    };
};

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {object} A map of each string
 */
export const toBoolMap = (array: string[]): BoolMap =>
    array.reduce(
        (map: BoolMap, item: string) => ({
            ...map,
            [item]: true
        }),
        {}
    );

/**
 * Inserts commas into numbers where appropriate for better readability
 * @param {number} value - A number, e.g. 10000
 * @returns {string} A comma-separated string, e.g. 10,000
 */
export const addCommas = (value: number): string =>
    value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export const snakify = (value: string): string =>
    value
        .toLowerCase()
        .trim()
        .replace(SNAKED_CHARS, '_');

/**
 * Returns a React component's name so we can attach it to a HOC's displayName property and view it in RDT
 * @param {ComponentClass | SFC} Component - A React component
 * @returns {string} The component's name
 */
export const getDisplayName = (WrappedComponent: ComponentClass | SFC): string =>
    WrappedComponent.displayName || WrappedComponent.name || 'Component';

/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
    componentWrapper: ReactWrapper<{}, {}> | ShallowWrapper<{}, {}>,
    specName: string,
    attributeName: string = 'data-spec'
): any => {
    return componentWrapper.find(`[${attributeName}="${specName}"]`);
};

/**
 * Returns true if a given UUID matches v4 format
 * @param {string} uuid - A version 4 UUID (no braces, uppercase OK)
 * @returns {boolean}
 */
export const validUUID = (uuid: string): boolean => V4_UUID.test(uuid);

/**
 * Returns a given string in title case, e.g. 'full name' becomes 'Full Name'
 * @param {string} str - string to be title-cased
 * @returns {string} Title-cased string
 */
export const titleCase = (str: string): string =>
    str.replace(/\b\w+/g, s => s.charAt(0).toUpperCase() + s.substr(1).toLowerCase());

export const getSelectClass = (errors: number): string => {
    if (errors === 0) {
        return '';
    }
    // We use a global selector here for react-select
    return 'select-invalid';
};

export const reorderList = (list: any[], startIndex: number, endIndex: number): any[] => {
    const [removed] = list.splice(startIndex, 1);

    list.splice(endIndex, 0, removed);

    return list;
};

/**
 * Compares basic objects (no methods and DOM nodes; property order important)
 * @param {object} objA - basic object
 * @param {object} objB - basic object
 * @returns {boolean}
 */
export const jsonEqual = (objA: {}, objB: {}): boolean =>
    JSON.stringify(objA) === JSON.stringify(objB);

/**
 * Checks whether any of a list of error strings contain one or more queries.
 * Used to determine whether a particular error has been encountered.
 */
export const hasErrorType = (errors: string[], exps: RegExp[]): boolean => {
    if (!errors.length) {
        return false;
    }
    for (const error of errors) {
        for (const exp of exps) {
            if (error.match(exp)) {
                return true;
            }
        }
    }
    return false;
};

export const pureSort = (list: any[], fn: (a: any, b: any) => number) => [...list].sort(fn);

export const getNodesBelow = ({ uuid: nodeUUID }: Node, nodes: Node[]) => {
    const idx = nodes.findIndex(({ uuid }: Node) => uuid === nodeUUID);
    return nodes.slice(idx, nodes.length);
};

export const getPendingConnection = (
    nodeUUID: string,
    pendingConnections: PendingConnections
): DragPoint => pendingConnections[nodeUUID];

export const getDetails = (uuid: string, components: Components) => components[uuid];

export const getNode = (uuid: string, components: Components, definition: FlowDefinition) => {
    const details = components[uuid];
    if (!details) {
        return null;
    }
    return definition.nodes[details.nodeIdx];
};

export const getExit = (uuid: string, components: Components, definition: FlowDefinition) => {
    const details = components[uuid];
    if (details) {
        const node = definition.nodes[details.nodeIdx];
        return node.exits[details.exitIdx];
    }
    return null;
};

export const getNodeUI = (uuid: string, definition: FlowDefinition) => definition._ui.nodes[uuid];

export const collides = (a: Bounds, b: Bounds) => {
    if (a.bottom < b.top || a.top > b.bottom || a.left > b.right || a.right < b.left) {
        return false;
    }
    // console.log("COLLISION!");
    return true;
};

export const getConnectionError = (source: string, targetUUID: string, components: Components) => {
    const { nodeUUID } = getDetails(source, components);
    if (nodeUUID === targetUUID) {
        return 'Connections cannot route back to the same places.';
    }
    return null;
};

/**
 * Computes translations prop for `Node` components in render()
 */
export const getTranslations = (definition: FlowDefinition, language: Language) => {
    if (definition.localization) {
        return definition.localization[language.iso];
    }
    return null;
};

export const getLocalizations = (
    node: Node,
    action: AnyAction,
    iso: string,
    languages: Languages,
    translations?: { [uuid: string]: any }
): LocalizedObject[] => {
    const localizations: LocalizedObject[] = [];

    // Account for localized cases
    if (node.router && node.router.type === 'switch') {
        const router = node.router as SwitchRouter;

        router.cases.forEach(kase =>
            localizations.push(Localization.translate(kase, iso, languages, translations))
        );

        // Account for localized exits
        if (node.exits) {
            node.exits.forEach(exit => {
                console.log('exit:', exit);
                localizations.push(Localization.translate(exit, iso, languages, translations));
            });
        }
    }

    if (action) {
        localizations.push(Localization.translate(action, iso, languages, translations));
    }

    return localizations;
};

export const determineConfigType = (
    nodeToEdit: Node,
    actions: AnyAction[],
    definition: FlowDefinition,
    components: Components
) => {
    if (actions.length) {
        return actions[actions.length - 1].type;
    } else {
        const nodeUI = definition._ui.nodes[nodeToEdit.uuid];
        if (nodeUI) {
            if (nodeUI.type) {
                return nodeUI.type;
            }
        }
    }

    // Account for ghost nodes
    if (nodeToEdit) {
        if (nodeToEdit.router) {
            return nodeToEdit.router.type;
        }

        if (nodeToEdit.actions) {
            return nodeToEdit.actions[0].type;
        }
    }

    const details = getDetails(nodeToEdit.uuid, components);
    if (details.type) {
        return details.type;
    }

    throw new Error(`Cannot initialize NodeEditor without a valid type: ${nodeToEdit.uuid}`);
};

export const getLocalizedObject = (localizations: LocalizedObject[]) => {
    if (localizations && localizations.length) {
        return localizations[0];
    }
};
