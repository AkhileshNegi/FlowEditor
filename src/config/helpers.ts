import { Operator, Type } from '~/config';
import { FlowTypes, FlowTypeVisibility } from '~/config/interfaces';

export const filterOperators = (operators: Operator[], flowType: FlowTypes): Operator[] => {
    // return filterVisibility(operators, flowType);
    return operators;
};

export const filterTypeConfigs = (typeConfigs: Type[], flowType: FlowTypes): Type[] => {
    return filterVisibility(typeConfigs, flowType);
};

const filterVisibility = (items: FlowTypeVisibility[], flowType: FlowTypes): any[] => {
    return items.filter((item: FlowTypeVisibility) => {
        if (item.visibility === undefined) {
            return true;
        }

        return item.visibility.findIndex((ft: FlowTypes) => ft === flowType) > -1;
    });
};
