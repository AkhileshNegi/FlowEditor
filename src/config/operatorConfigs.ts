import { HIDDEN, Operator, OperatorMap, Operators, TEXT_TYPES } from '~/config/interfaces';

export const operatorConfigList: Operator[] = [
    {
        type: Operators.has_any_word,
        verboseName: 'has any of the words',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_all_words,
        verboseName: 'has all of the words',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_phrase,
        verboseName: 'has the phrase',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_only_phrase,
        verboseName: 'has only the phrase',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_beginning,
        verboseName: 'starts with',
        operands: 1
    },
    {
        type: Operators.has_text,
        verboseName: 'has some text',
        operands: 0,
        categoryName: 'Has Text',
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_number,
        verboseName: 'has a number',
        operands: 0,
        categoryName: 'Has Number'
    },
    {
        type: Operators.has_number_between,
        verboseName: 'has a number between',
        operands: 2
    },
    {
        type: Operators.has_number_lt,
        verboseName: 'has a number below',
        operands: 1
    },
    {
        type: Operators.has_number_lte,
        verboseName: 'has a number at or below',
        operands: 1
    },
    {
        type: Operators.has_number_eq,
        verboseName: 'has a number equal to',
        operands: 1
    },
    {
        type: Operators.has_number_gte,
        verboseName: 'has a number at or above',
        operands: 1
    },
    {
        type: Operators.has_number_gt,
        verboseName: 'has a number above',
        operands: 1
    },
    {
        type: Operators.has_date,
        verboseName: 'has a date',
        operands: 0,
        categoryName: 'Has Date',
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_date_lt,
        verboseName: 'has a date before',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_date_eq,
        verboseName: 'has a date equal to',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_date_gt,
        verboseName: 'has a date after',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_run_status,
        verboseName: 'has a run status of',
        operands: 1,
        visibility: HIDDEN
    },
    {
        type: Operators.has_group,
        verboseName: 'is in the group',
        operands: 1,
        visibility: HIDDEN
    },
    {
        type: Operators.has_phone,
        verboseName: 'has a phone number',
        operands: 0,
        categoryName: 'Has Phone'
    },
    {
        type: Operators.has_email,
        verboseName: 'has an email',
        operands: 0,
        categoryName: 'Has Email',
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_error,
        verboseName: 'has an error',
        operands: 0,
        categoryName: 'Has Error',
        visibility: HIDDEN
    },
    {
        type: Operators.has_value,
        verboseName: 'is not empty',
        operands: 0,
        categoryName: 'Not Empty'
    },
    {
        type: Operators.has_wait_timed_out,
        verboseName: null,
        operands: 0,
        visibility: HIDDEN
    },
    {
        type: Operators.has_pattern,
        verboseName: 'matches regex',
        operands: 1
    }
];

export const operatorConfigMap: OperatorMap = operatorConfigList.reduce(
    (map: OperatorMap, operatorConfig: Operator) => {
        map[operatorConfig.type] = operatorConfig;
        return map;
    },
    {}
);

/**
 * Shortcut for constant lookup of operator config in operator configs map
 * @param {string} type - The type of the operator config to return, e.g. 'send_msg'
 * @returns {Object} - The operator config found at operatorConfigs[type] or -1
 */
export const getOperatorConfig = (type: Operators): Operator => operatorConfigMap[type];
