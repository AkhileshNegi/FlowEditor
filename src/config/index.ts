import ConfigProvider, { ConfigProviderContext } from './ConfigProvider';
import { Type, Mode, typeConfigList, getTypeConfig } from './typeConfigs';
import { Operator, operatorConfigList, getOperatorConfig } from './operatorConfigs';

export {
    ConfigProviderContext,
    Type,
    Mode,
    Operator,
    typeConfigList,
    getTypeConfig,
    operatorConfigList,
    getOperatorConfig
};

export default ConfigProvider;
