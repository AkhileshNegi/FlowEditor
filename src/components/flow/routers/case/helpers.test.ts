import { operatorConfigList, Operators } from '~/config/operatorConfigs';
import {
    prefix,
    composeExitName,
    getMinMax,
    isFloat,
    isInt,
    strContainsNum,
    parseNum
} from '~/components/flow/routers/case/helpers';
import { titleCase } from '~/utils';

describe('helpers', () => {
    describe('prefix', () =>
        operatorConfigList.forEach(({ verboseName, type }) =>
            it(`should prefix "${verboseName}" operator appropriately`, () => {
                if (verboseName) {
                    expect(prefix(type)).toMatchSnapshot();
                }
            })
        ));

    describe('composeExitName', () => {
        const operatorType = Operators.has_any_word;
        const strArgOperators = operatorConfigList.slice(0, 6);
        let args = [['tomato, t'], ['papayawhip'], []];

        it('should handle empty arg lists appropriately', () =>
            expect(composeExitName(operatorType, args[2], '')).toBe(''));

        it('should return the first arg in list, capitalized', () => {
            args = args.slice(0, 2);
            strArgOperators.forEach(({ type }) =>
                args.slice(0, 2).forEach(argList => {
                    const [firstArg] = argList[0].split(',');
                    expect(composeExitName(type, argList, '')).toBe(titleCase(firstArg));
                })
            );
        });

        it('should return exit name in the format "min - max" if operator', () => {
            [['1', '2'], ['', '2'], ['1', ''], []].forEach(argList => {
                expect(
                    composeExitName(Operators.has_number_between, argList, '').indexOf('-') > -1
                ).toBeTruthy();
            });
        });

        it("should return newExitName if it's truthy and operator is 'has_number_between'", () => {
            expect(composeExitName(Operators.has_number_between, ['1', '2'], 'Violet')).toBe(
                'Violet'
            );
        });
    });

    describe('getMinMax', () => {
        const emptyStrs = { min: '', max: '' };

        it('should return emtpy strings if arg param is an empty array', () => {
            expect(getMinMax([])).toEqual(emptyStrs);
        });

        it("should return empty strings if arg param doesn't contain two number-containing strings", () => {
            expect(getMinMax(['a', 'b'])).toEqual(emptyStrs);
        });

        it('should return a number-containing string MIN and empty string MAX if first arg is a number-containing string and second arg is not', () => {
            expect(getMinMax(['1', 'b'])).toEqual({ min: '1', max: '' });
        });

        it('should return an empty string MIN and number-containing string MAX if first arg is not a number-containing string and second arg is', () => {
            expect(getMinMax(['a', '2'])).toEqual({ min: '', max: '2' });
        });
    });

    describe('isFloat', () => {
        it('should return true if argument is string containing a float', () => {
            ['0.2', '.2', '+.2', '+0.2', '-.2', '-0.2', '2', '2.'].forEach(arg =>
                expect(isFloat(arg as any)).toBeTruthy()
            );
        });

        it('should return false if argument is not a string containing a float', () => {
            ['u.2', '0.2u', 'a'].forEach(arg => expect(isFloat(arg as any)).toBeFalsy());
        });
    });

    describe('isInt', () => {
        it('should return true if argument is string containing an int', () => {
            ['1', '+1', '-1'].forEach(arg => expect(isInt(arg)).toBeTruthy());
        });

        it('should return false if argument is not a string containing an int', () => {
            ['0.1', 'e24', '-.3', '5+', '5-', 'a'].forEach(arg => expect(isInt(arg)).toBeFalsy());
        });
    });

    describe('strContainsNum', () => {
        it('should return true if string contains only a float or int', () => {
            ['0.2', '.2', '+.2', '+0.2', '-.2', '-0.2', '2', '2.', '1', '+1', '-1'].forEach(arg =>
                expect(strContainsNum(arg)).toBeTruthy()
            );
        });

        it('should return false if string does not contain only a float or int', () => {
            ['a', 'e24', '2a', '0.2u', 'u.2', 'u0.2'].forEach(arg =>
                expect(strContainsNum(arg)).toBeFalsy()
            );
        });
    });

    describe('parseNum', () => {
        it('should return a float if passed a string containing only a float', () => {
            const float = parseNum('0.2');

            expect(typeof float === 'number' && isFinite(float)).toBeTruthy();
        });

        it('should return an int if passed a string containing only an int', () => {
            const int = parseNum('1');

            expect(typeof int === 'number' && isFinite(int)).toBeTruthy();
        });
    });
});
