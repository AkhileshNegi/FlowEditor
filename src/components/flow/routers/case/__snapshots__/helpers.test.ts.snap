// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers prefix should prefix "has a date after" operator appropriately 1`] = `"After "`;

exports[`helpers prefix should prefix "has a date before" operator appropriately 1`] = `"Before "`;

exports[`helpers prefix should prefix "has a date equal to" operator appropriately 1`] = `""`;

exports[`helpers prefix should prefix "has a date" operator appropriately 1`] = `""`;

exports[`helpers prefix should prefix "has a number above" operator appropriately 1`] = `">"`;

exports[`helpers prefix should prefix "has a number at or above" operator appropriately 1`] = `">= "`;

exports[`helpers prefix should prefix "has a number at or below" operator appropriately 1`] = `"<= "`;

exports[`helpers prefix should prefix "has a number below" operator appropriately 1`] = `"< "`;

exports[`helpers prefix should prefix "has a number between" operator appropriately 1`] = `""`;

exports[`helpers prefix should prefix "has a number equal to" operator appropriately 1`] = `""`;

exports[`helpers prefix should prefix "has a number" operator appropriately 1`] = `""`;

exports[`helpers prefix should prefix "has a phone number" operator appropriately 1`] = `""`;

exports[`helpers prefix should prefix "has a run status of" operator appropriately 1`] = `""`;

exports[`helpers prefix should prefix "has all of the words" operator appropriately 1`] = `""`;

exports[`helpers prefix should prefix "has an email" operator appropriately 1`] = `""`;

exports[`helpers prefix should prefix "has an error" operator appropriately 1`] = `""`;

exports[`helpers prefix should prefix "has any of the words" operator appropriately 1`] = `""`;

exports[`helpers prefix should prefix "has only the phrase" operator appropriately 1`] = `""`;

exports[`helpers prefix should prefix "has some text" operator appropriately 1`] = `""`;

exports[`helpers prefix should prefix "has the phrase" operator appropriately 1`] = `""`;

exports[`helpers prefix should prefix "is in the group" operator appropriately 1`] = `""`;

exports[`helpers prefix should prefix "is not empty" operator appropriately 1`] = `""`;

exports[`helpers prefix should prefix "starts with" operator appropriately 1`] = `""`;

exports[`helpers validateCase cross validates min - max 1`] = `
Object {
  "argument": Object {
    "value": "",
  },
  "exitName": Object {
    "value": "50 - 3",
  },
  "exitNameEdited": false,
  "max": Object {
    "validationFailures": Array [
      Object {
        "message": "Maximum value must be a more than the minimum",
      },
    ],
    "value": "3",
  },
  "min": Object {
    "validationFailures": Array [
      Object {
        "message": "Minimum value must be a less than the maximum",
      },
    ],
    "value": "50",
  },
  "operatorConfig": Object {
    "operands": 2,
    "type": "has_number_between",
    "verboseName": "has a number between",
  },
}
`;

exports[`helpers validateCase does not require arguments without an exit 1`] = `
Object {
  "argument": Object {
    "value": "",
  },
  "exitName": Object {
    "value": "",
  },
  "exitNameEdited": false,
  "max": Object {
    "value": "",
  },
  "min": Object {
    "value": "",
  },
  "operatorConfig": Object {
    "operands": 1,
    "type": "has_any_word",
    "verboseName": "has any of the words",
  },
}
`;

exports[`helpers validateCase does not suggest an empty range for exit name, ie " - " 1`] = `
Object {
  "argument": Object {
    "value": "",
  },
  "exitName": Object {
    "value": "",
  },
  "exitNameEdited": false,
  "max": Object {
    "value": "",
  },
  "min": Object {
    "value": "",
  },
  "operatorConfig": Object {
    "operands": 2,
    "type": "has_number_between",
    "verboseName": "has a number between",
  },
}
`;

exports[`helpers validateCase requires arguments when exit is set 1`] = `
Object {
  "argument": Object {
    "validationFailures": Array [
      Object {
        "message": "Value is required",
      },
    ],
    "value": "",
  },
  "exitName": Object {
    "value": "My Exit",
  },
  "exitNameEdited": true,
  "max": Object {
    "value": "",
  },
  "min": Object {
    "value": "",
  },
  "operatorConfig": Object {
    "operands": 1,
    "type": "has_any_word",
    "verboseName": "has any of the words",
  },
}
`;

exports[`helpers validateCase suggests min - max exits 1`] = `
Object {
  "argument": Object {
    "value": "",
  },
  "exitName": Object {
    "value": "5 - 40",
  },
  "exitNameEdited": false,
  "max": Object {
    "value": "40",
  },
  "min": Object {
    "value": "5",
  },
  "operatorConfig": Object {
    "operands": 2,
    "type": "has_number_between",
    "verboseName": "has a number between",
  },
}
`;

exports[`helpers validateCase suggests operator exit names for no operands 1`] = `
Object {
  "argument": Object {
    "value": "",
  },
  "exitName": Object {
    "value": "Has Text",
  },
  "exitNameEdited": false,
  "max": Object {
    "value": "",
  },
  "min": Object {
    "value": "",
  },
  "operatorConfig": Object {
    "categoryName": "Has Text",
    "operands": 0,
    "type": "has_text",
    "verboseName": "has some text",
  },
}
`;
